ksql-streams/src/main/java/io/confluent/ksql/execution/streams/materialization/ks/KsLocator.java

```diff
@@ -51,20 +60,51 @@
   }
 
   @Override
-  @SuppressWarnings("deprecation")
-  public Optional<KsqlNode> locate(final Struct key) {
-    final StreamsMetadata metadata = kafkaStreams
-        .metadataForKey(stateStoreName, key, keySerializer);
+  public Optional<ActiveAndStandByNodes> locate(
+      final Struct key,
+      final Optional<Map<String, HostStatusEntity>> hostStatuses
+  ) {
+    final KeyQueryMetadata metadata = kafkaStreams
+        .queryMetadataForKey(stateStoreName, key, keySerializer);
 
-    if (metadata == StreamsMetadata.NOT_AVAILABLE) {
+    // FAil fast if Streams not ready. Let client handle it
+    if (metadata == KeyQueryMetadata.NOT_AVAILABLE) {
+      LOG.debug("Streams Metadata not available");
       return Optional.empty();
     }
 
-    final HostInfo hostInfo = metadata.hostInfo();
-    return Optional.of(asNode(hostInfo));
+    final HostInfo activeHost = metadata.getActiveHost();
+    final KsqlNode activeNode = asNode(activeHost);
+    final Set<HostInfo> standByHosts = metadata.getStandbyHosts();
+    final Map<HostInfo, KsqlNode> standByNodesMap = new HashMap<>();
+    for (HostInfo hostInfo: standByHosts) {
+      standByNodesMap.put(hostInfo, asNode(hostInfo));
+    }
+
+    LOG.debug("Active host {} , standby hosts {}", activeHost, standByHosts);
```

Comment:
by marking these statuses.. this interface/class is doing more than "locating", right?

Reply:
Refactored so now it is locating and filtering. Whatever is returned from the locator is assumed to be a valid node (alive and less than max allowed lag)